### PR TITLE
[NTUSER] Fix sticky mouse buttons

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Fix Mouse Buttons Sticking Down causing unexpected window dragging._
_This is a patch supplied by @I_Kill_Bugs and seems to work well as tested by @julenuri._

JIRA issue: [CORE-11775](https://jira.reactos.org/browse/CORE-11775)
JIRA issue: [CORE-14998](https://jira.reactos.org/browse/CORE-14998)
Improves JIRA issue: [CORE-18511](https://jira.reactos.org/browse/CORE-18511)

## Proposed changes

_Add extra code into nonclient.c procedure DefWndDoSizeMove to drop tracking._
_If we get a mouse move with the mouse left button down, then break out of testing._

### CODE REVIEW

File: win32ss/user/ntuser/nonclient.c [hunk 1]
```diff
@@ -394,11 +394,13 @@ DefWndDoSizeMove(PWND pwnd, WORD wParam)
       if (msg.message == WM_KEYDOWN && (msg.wParam == VK_RETURN || msg.wParam == VK_ESCAPE))
          break; // Exit on Return or Esc
 
-      if (!g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP))
+      if (!g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP ||
+         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) != MK_LBUTTON)))
```
Comment (SergeGautherie) [inline win32ss/user/ntuser/nonclient.c:398]:
```suggestion
      if (!g_bWindowSnapEnabled &&
          (msg.message == WM_LBUTTONUP ||
           (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) != MK_LBUTTON)))
```

File: win32ss/user/ntuser/nonclient.c [hunk 2]
```diff
@@ -394,11 +394,13 @@ DefWndDoSizeMove(PWND pwnd, WORD wParam)
...
       { // If no WindowSnapEnabled: Exit on button-up immediately
          break;
       }
-      else if (g_bWindowSnapEnabled && msg.message == WM_LBUTTONUP)
+      else if (g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP ||
+         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) != MK_LBUTTON)))
```
Comment (SergeGautherie) [inline win32ss/user/ntuser/nonclient.c:403]:
```suggestion
      else if (g_bWindowSnapEnabled &&
               (msg.message == WM_LBUTTONUP ||
                (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) != MK_LBUTTON)))
```

File: win32ss/user/ntuser/nonclient.c (hunk 1 above)
Comment (Doug-Lyons) [reply win32ss/user/ntuser/nonclient.c:398]: In the original layout the if test line break is on the only "||" which makes better sense to me. Thanks.

File: win32ss/user/ntuser/nonclient.c (hunk 2 above)
Comment (Doug-Lyons) [reply win32ss/user/ntuser/nonclient.c:403]: In the original layout the if test line break is on the only "||" which makes better sense to me. Thanks.

File: win32ss/user/ntuser/nonclient.c (hunk 1 above)
Comment (binarymaster) [inline win32ss/user/ntuser/nonclient.c:398]:
```suggestion
         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) == 0)))
```

File: win32ss/user/ntuser/nonclient.c (hunk 2 above)
Comment (binarymaster) [inline win32ss/user/ntuser/nonclient.c:403]:
```suggestion
         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) == 0)))
```

**Review by binarymaster — COMMENTED** [review]
Essentially `(int & BIT_MASK) != BIT_MASK` pattern is the same as `(int & BIT_MASK) == 0`, so please simplify that.  Also, do not capitalize each word in PR title.  Module prefix also should be `[NTUSER]`.

**Comment by julenuri** [comment]
Be careful about the cascade effect that appears after. I know that it might be another bug, that is related to them, but it should be investigated after applying this. 

Can this code (or the code before) be simplified to just drag and drop it correctly without the exceptions? Is this Windows aware/like? Win32k  should be the less hacky as possible. :100:

File: win32ss/user/ntuser/nonclient.c (hunk 1 above)
Comment (Doug-Lyons) [reply win32ss/user/ntuser/nonclient.c]: Done. Thanks.

File: win32ss/user/ntuser/nonclient.c (hunk 2 above)
Comment (Doug-Lyons) [reply win32ss/user/ntuser/nonclient.c]: Done. Thanks.

**Comment by HBelusca** [comment]
> [...] but it should be investigated after applying this.

@julenuri What would be the best course of action with this PR then? Wait for merging it until that other problem becomes understood and fixed as well? or?

File: win32ss/user/ntuser/nonclient.c [hunk 3]
```diff
@@ -394,11 +394,13 @@ DefWndDoSizeMove(PWND pwnd, WORD wParam)
...
 
-      if (!g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP))
+      if (!g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP ||
+         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) == 0)))
       { // If no WindowSnapEnabled: Exit on button-up immediately
          break;
       }
-      else if (g_bWindowSnapEnabled && msg.message == WM_LBUTTONUP)
+      else if (g_bWindowSnapEnabled && (msg.message == WM_LBUTTONUP ||
+         (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) == 0)))
```
Comment (HBelusca) [inline win32ss/user/ntuser/nonclient.c:403]:
```suggestion
              (msg.message == WM_MOUSEMOVE && (msg.wParam & MK_LBUTTON) == 0)))
```

**Review by HBelusca — COMMENTED** [review]
Minor formatting stuff

File: win32ss/user/ntuser/nonclient.c (hunk 3 above)
Comment (Doug-Lyons) [reply win32ss/user/ntuser/nonclient.c]: Fixed. Thanks.

**Comment by julenuri** [comment]
> > [...] but it should be investigated after applying this.
> 
> @julenuri What would be the best course of action with this PR then? Wait for merging it until that other problem becomes understood and fixed as well? or?

In my opinion, first merge (and think about if it is worthy to retro apply into the releases, I do think so). 


> Is there any improvement also when testing both this PR with the other #6427 together ?

I haven't tested them combined. I would like to do also summed to the acceleration under real hardware to check if this would fix some bad draws.

**Review by JoachimHenze — APPROVED** [review]
Approved, as the promise of this PR is tempting.
Most likely it is a workaround though.

>> In my opinion, first merge (and think about if it is worthy to retro apply into the releases, I do think so).

I generally don't rush anything with backporting NTUSER-fixes, and especially not if they are looking like workarounds. So this will not happen until I have observed its effect on master for half a year at least.

**Comment by Doug-Lyons** [comment]
Thanks for approving and committing this, @JoachimHenze.

**Comment by binarymaster** [comment]
This PR introduced regression: https://jira.reactos.org/browse/CORE-19503

**Pull request merged**